internal/impl/redpanda/processor_data_transforms.go

```diff
@@ -0,0 +1,414 @@
+	ctx := context.Background()
+	runtimeCfg := wazero.NewRuntimeConfig().
+		WithCloseOnContextDone(true).
+		WithCompilationCache(wazero.NewCompilationCache()).
+		WithMemoryLimitPages(uint32(cfg.maxMemoryPages))
+	r := wazero.NewRuntimeWithConfig(ctx, runtimeCfg)
+	cm, err := r.CompileModule(ctx, wasmBinary)
+	_ = r.Close(ctx)
+	if err != nil {
+		return nil, err
+	}
+	// TODO: Validate ABI contract
+	proc := &dataTransformEnginePool{
+		log:           mgr.Logger(),
+		modulePool:    sync.Pool{},
+		runtimeConfig: runtimeCfg,
+		wasmBinary:    cm,
+		cfg:           cfg,
+	}
+	// Ensure we can create at least one module runner.
+	modRunner, err := proc.newModule()
+	if err != nil {
+		return nil, err
+	}
+
+	proc.modulePool.Put(modRunner)
+	return proc, nil
+}
+
+func (p *dataTransformEnginePool) newModule() (engine *dataTransformEngine, err error) {
+	ctx := context.Background()
+	r := wazero.NewRuntimeWithConfig(ctx, p.runtimeConfig)
+	engine = &dataTransformEngine{
+		log:       p.log,
+		cfg:       &p.cfg,
+		runtime:   r,
+		hostChan:  make(chan any),
+		guestChan: make(chan any),
+		procErr:   nil,
+	}
+	defer func() {
+		if err != nil {
+			engine.runtime.Close(context.Background())
+		}
+	}()
+
+	builder := r.NewHostModuleBuilder("redpanda_transform")
+	for name, ctor := range transformHostFunctions {
+		builder = builder.NewFunctionBuilder().WithFunc(ctor(engine)).Export(name)
+	}
+	if _, err = builder.Instantiate(ctx); err != nil {
+		return
+	}
+
+	if _, err = wasi_snapshot_preview1.Instantiate(ctx, r); err != nil {
+		return
+	}
+	cfg := wazero.NewModuleConfig().
+		WithStartFunctions().
+		WithArgs("transform").
+		WithName("transform").
+		WithEnv("REDPANDA_INPUT_TOPIC", "benthos")
+	for i := 0; i < 8; i += 1 {
+		cfg = cfg.WithEnv(fmt.Sprintf("REDPANDA_OUTPUT_TOPIC_%d", i), fmt.Sprintf("output_%d", i))
+	}
+	if engine.mod, err = r.InstantiateModule(ctx, p.wasmBinary, cfg); err != nil {
+		return
+	}
+	start := engine.mod.ExportedFunction("_start")
+	if start == nil {
+		err = errors.New("_start function is required")
+		engine.mod.Close(ctx)
+		return
+	}
+	go func() {
+		_, err := start.Call(context.Background())
+		if !engine.mod.IsClosed() {
+			_ = engine.mod.Close(context.Background())
+		}
+		if err == nil {
+			err = sys.NewExitError(0)
+		}
+		engine.procErr = err
+		close(engine.hostChan)
+	}()
+
+	// Wait for the engine to start
+	select {
+	case <-engine.hostChan:
+	case <-time.After(p.cfg.timeout):
+		_ = engine.mod.Close(ctx)
+		drainChannel(engine.hostChan) // Wait for goroutine to exit
+	}
+	return engine, engine.procErr
+}
+
+func (p *dataTransformEnginePool) ProcessBatch(ctx context.Context, batch service.MessageBatch) ([]service.MessageBatch, error) {
+	var modRunner *dataTransformEngine
+	var err error
+	if modRunnerPtr := p.modulePool.Get(); modRunnerPtr != nil {
+		modRunner = modRunnerPtr.(*dataTransformEngine)
+	} else {
+		if modRunner, err = p.newModule(); err != nil {
+			return nil, err
+		}
+	}
```

Comment:
I wonder if this could be simplified by setting the `New` func in `sync.Pool`: https://pkg.go.dev/sync#Pool

Reply:
sync.Pool#New does not allow for error reporting, which is why this pattern is here.

Reply:
Good point, thank you for the explanation!